pubsub/google/cloud/pubsub_v1/subscriber/policy/base.py

```diff
@@ -0,0 +1,273 @@
+                data for predicting appropriate ack times. If set, this should
+                be a dictionary-like object.
+
+                .. note::
+                    Additionally, the histogram relies on the assumption
+                    that the dictionary will properly sort keys provided
+                    that all keys are positive integers. If you are sending
+                    your own dictionary class, ensure this assumption holds
+                    or you will get strange behavior.
+        """
+        self._client = client
+        self._subscription = subscription
+        self._consumer = consumer.Consumer(self)
+        self._ack_deadline = 10
+        self._last_histogram_size = 0
+        self.histogram = histogram.Histogram(data=histogram_data)
+
+    @property
+    def ack_deadline(self):
+        """Return the appropriate ack deadline.
+
+        This method is "sticky". It will only perform the computations to
+        check on the right ack deadline if the histogram has gained a
+        significant amount of new information.
+
+        Returns:
+            int: The correct ack deadline.
+        """
+        if len(self.histogram) > self._last_histogram_size * 2:
+            self._ack_deadline = self.histogram.percentile(percent=99)
+        return self._ack_deadline
+
+    @property
+    def initial_request(self):
+        """Return the initial request.
+
+        This defines the initial request that must always be sent to Pub/Sub
+        immediately upon opening the subscription.
+        """
+        return types.StreamingPullRequest(
+            stream_ack_deadline_seconds=self.histogram.percentile(99),
+            subscription=self.subscription,
+        )
+
+    @property
+    def managed_ack_ids(self):
+        """Return the ack IDs currently being managed by the policy.
+
+        Returns:
+            set: The set of ack IDs being managed.
+        """
+        if not hasattr(self, '_managed_ack_ids'):
+            self._managed_ack_ids = set()
+        return self._managed_ack_ids
+
+    @property
+    def subscription(self):
+        """Return the subscription.
+
+        Returns:
+            str: The subscription
+        """
+        return self._subscription
+
+    def ack(self, ack_id):
+        """Acknowledge the message corresponding to the given ack_id.
+
+        Args:
+            ack_id (str): The ack ID.
+        """
+        request = types.StreamingPullRequest(ack_ids=[ack_id])
+        self._consumer.send_request(request)
+
+    def call_rpc(self, request_generator):
+        """Invoke the Pub/Sub streaming pull RPC.
+
+        Args:
+            request_generator (Generator): A generator that yields requests,
+                and blocks if there are no outstanding requests (until such
+                time as there are).
+        """
+        return self._client.api.streaming_pull(request_generator,
+            options=gax.CallOptions(timeout=600),
```

Comment:
Why 600? The StreamingPull connection will be good for a half of an hour, and if this is a timeout in case of an inability to connect this seems too long.